# Data Cleaning with Pandas



In [59]:
import pandas as pd

In [60]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

In [61]:
orders_df = orders.copy()

In [62]:
orderlines_df = orderlines.copy()

## 1.&nbsp; Duplicates


In [63]:
# orders
orders_df.duplicated().sum()

0

In [64]:
# orderlines
orderlines_df.duplicated().sum()

0

# 2.&nbsp; `.info()`

In [65]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [66]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [67]:
print(f"5 missing values represents {((orders_df.total_paid.isna().sum() / orders_df.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


In [68]:
orders_df.total_paid.isna().value_counts(normalize=True)

total_paid
False    0.999978
True     0.000022
Name: proportion, dtype: float64

As there is such a tiny amount of missing values, I will simply delete these rows, as I have enough data without them.

In [69]:
orders_df = orders_df.loc[~orders.total_paid.isna(), :]

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [70]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])

<ipython-input-70-87c26cc5d8b0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [71]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])

#### 4.1.2.&nbsp;`unit_price`

In [72]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: Unable to parse string "1.137.99" at position 6

 When I try to convert `unit_price` to a numerical datatype, I receive a `ValueError` telling me that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.*kursiver Text*

In [74]:
orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

unit_price
False    257814
True      36169
Name: count, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [75]:
two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


<ipython-input-75-d7cb20f99e43>:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and I might even end up losing a larger portion of the data than this too. For the moment I will delete the rows as I only have 2 weeks for this project and I'd like some quick, accurate results to show.

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), I need to remove the whole order and not just the affected row. If I only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

I therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [76]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]

orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

In [77]:
orderlines_df.shape[0]

216250

I still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [78]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

<ipython-input-78-d6bd38092a9f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])


It worked perfectly

#Clean the `products` DataFrame


In [79]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)

In [80]:
products_df = products.copy()

In [81]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

## Duplicates

In [82]:
products_df.duplicated().sum()

8746

Let's get rid of the duplicates

In [83]:
products_df = products_df.drop_duplicates()

In [84]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[10580 rows x 7 columns]

## `.info()`

In [85]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


### Missing values
You can see from `.info()` above that I have missing values in `desc` and `price`

#### `desc`

In [86]:
products_df["desc"].isna().sum()

7

7 is a very small number to have missing, let's have a closer look

In [87]:
products_df.loc[products_df['desc'].isna(), :]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

The products names here are quite descriptive, so I'm tempted to just copy them to the description column, so that there is a description if I later want utilise this column. I wouldn't recommend this if this DataFrame was the source of truth for the website. But this is not the case here, and I'm not faking any information (guessing a price or so), so I'm happy with this option

#### `price`

In [88]:
products_df.price.isna().sum()

46

In [89]:
print(f"The missing values in price are {(products_df.price.isna().value_counts(normalize=True)[1] * 100).round(2)}% of all rows in the DataFrame")

The missing values in price are 0.43% of all rows in the DataFrame


<ipython-input-89-149a8976a30c>:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"The missing values in price are {(products_df.price.isna().value_counts(normalize=True)[1] * 100).round(2)}% of all rows in the DataFrame")


Let's simply delete these rows to ensure that we can trust the numbers in our final DataFrame. Afterall, the price is very important when investigating discounts.

Option 1: `.loc`

In [90]:
products_df = products_df.loc[~products['price'].isna()]

In [91]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10534 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10534 non-null  object
 1   name         10534 non-null  object
 2   desc         10527 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10534 non-null  object
 5   in_stock     10534 non-null  int64 
 6   type         10484 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.4+ KB


#### `price`

First, let's see how many values are affected by the 2-decimal-dot problems or 3 decimal places.

In [92]:
price_problems_number = products_df.loc[(products_df.price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
price_problems_number

542

In [93]:
print(f"The column price has in total {price_problems_number} wrong values. This is {round(((price_problems_number / products_df.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column price has in total 542 wrong values. This is 5.15% of the rows of the DataFrame


5.15% is a reasonable amount of our data. However, the price column will be important to understanding discounts, so I'd like it to be very trustworthy as I am  basing business decisions on it. Therefore, I will delete these rows

In [94]:
products_df = products_df.loc[(~products_df.price.astype(str).str.contains("\d+\.\d+\.\d+"))&(~products_df.price.astype(str).str.contains("\d+\.\d{3,}")), :]

In [95]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                    desc  price promo_price  \
0           Aluminum support compatible with all MacBook  59.99     499.899   
1              USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                                 mouse Apple USB cable.     59     569.898   
3                  IPhone dock and USB Cable Apple iPod.     25     229.997   
4      2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   
...                                                  ...    ...         ...   
19321  compact and portable stand vertically or horiz...  29.99     269.903   
19322  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19323  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19324  Backpack with capacity of 14 liter compartment...  69.95     649.903   
19325  Backpack with capacity of 14 liter compartment...  69.95     649.903   

       in_stock      type  
0             1      8696  
1             0  13855401  
2             0      1387  
3             0      1230  
4             1      1364  
...         ...       ...  
19321         1     12282  
19322         1      1392  
19323         1      1392  
19324         0      1392  
19325         1      1392  

[9992 rows x 7 columns]

To complete our task, let's convert the column to a numeric datatype

In [96]:
products_df["price"] = pd.to_numeric(products_df["price"])

In [97]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9992 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          9992 non-null   object 
 1   name         9992 non-null   object 
 2   desc         9986 non-null   object 
 3   price        9992 non-null   float64
 4   promo_price  9992 non-null   object 
 5   in_stock     9992 non-null   int64  
 6   type         9946 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 624.5+ KB


#### `promo_price`

Again, let's begin by seeing how many values are affected by the 2-decimal-dots problem, or the 3 decimal-places problem

In [98]:
promo_problems_number = products_df.loc[(products_df.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :].shape[0]
promo_problems_number

9232

In [99]:
print(f"The column promo_price has in total {promo_problems_number} wrong values. This is {round(((promo_problems_number / products_df.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9232 wrong values. This is 92.39% of the rows of the DataFrame


In [100]:
promo_price_df = products_df.loc[(products_df.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :]
promo_price_df.sample(50)

sku                                               name  \
11334    TRN0011       MacBook sleeve Sleeve Trunk / Air 13 "Silver   
13426    APP1809                     Apple Apple Pack 4 Pencil Tips   
13172    PAC1507  Pack QNAP TS-251A NAS Server | 2GB RAM | WD 8T...   
1881   CRU0037-4  Crucial Mac Memory 32GB (4x8GB) 1866MHz DDR3 DIMM   
15207    ELA0022       elago Lightning Cable 1m Aluminum Space Gray   
9114     APP1257  Apple iMac 21.5 "Core i5 2.8GHz | 8GB | 1TB Fu...   
12963  BNQ0039-A          (Open) LED Monitor Benq GW2470H 238 "HDMI   
19188    XRI0011                   X-Rite color calibrator i1Studio   
16846    GLY0011  Glyph Atom SSD SSD RAID 2TB USB-C (3.1Gen2) Si...   
15921    PAC2311  Synology DS1517 + NAS Server | 2GB RAM | 60TB ...   
12574    IFX0070  iFixit part adhesive strips Battery iPhone 6 Plus   
2564     TUC0203  Tucano MacBook Pro Backpack Livello Up / Retin...   
18008    MOS0241    Moshi iPhone Case StealthCover X Rosa Champagne   
17595    OWC0236  OWC USB Travel Dock-C / USB 3.1 / HDMI / SD Ca...   
12964  TRA0017-A  Open - Transcend JetDrive SSD 720 240GB Macboo...   
15020    BEL0300  Belkin Rockstar Lightning Audio Adapter + Char...   
16663    APP2142  Apple iPad Pro 10.5 "Wi-Fi + Cellular 512GB Si...   
516      SEV0007             RAM installation service MacBook / Pro   
354      APP0431                          Apple iPod nano 16Gb Blue   
15475    REP0386                 Repair iPad Mini 3 Headphone Input   
11899    QDO0012  QDOS OptiGuard Titanium protector iPhone 6s Pl...   
17860    PAC2228  Synology DS218 + NAS Server | 2GB RAM | 8TB (2...   
2548     BOS0019                   Bose SoundLink Color Red Speaker   
15082  APP0271-A  Open - Apple Wireless Keyboard Keyboard (OEM) Mac   
17766    PAC2398  DS418play Synology NAS Server | 6GB RAM | 32TB...   
2609     DEN0011                         Denon Heos 5 Black Speaker   
17267    AP20278  Like new - Apple Watch 42mm steel case and str...   
687      PAC0494  Apple Mac Pro Quad-Core 37GHz | 64GB RAM | 512...   
13619    APP1801  Apple Correa 42mm Braiding Nailon Pale Pink / ...   
12944    SPH0015                  Sphero Robot Ollie Darkside Black   
13418    APP1697       Apple iPhone Leather Case Cover Black Plus 7   
13586    APP1726  Apple Watch Series 2 42mm Stainless Steel Case...   
15709    STC0009   Stacked Wall Charger Wireless Charger White wall   
11892  LOG0144-A  (Open) Logitech + iPhone support clamp Univers...   
13079    GTE0066  G-Technology G-DRIVE mobile USB 3.0 500GB FW80...   
19032    OWC0278         ThunderBay OWC Dual Thunderbolt RAID 5 4 3   
13024    PAC1489  Samsung SSD 850 expansion kit EVO 500 GB to 21...   
17317    WDT0392  WD My Passport 1TB USB External Hard Drive 25 ...   
1707     BEL0172    Belkin Slim-Fit Armband Plus iPhone 6 / 6S Rosa   
10737    ICA0056   Incase Systm Smart Case Cover iPhone 6 / 6S Gray   
11654    UAG0034          UAG Ash MacBook Pro Retina Case 13 "Black   
5613     PAC1055  Apple iMac 27 "Core i7 Retina 5K 4GHz | 32GB |...   
9153     APP1260  Apple iMac 21.5 "Core i5 2.8GHz | 16GB | 2TB F...   
2847     HOM0007          Homido Virtual Reality Headset for iPhone   
13594    APP1732  Apple Watch Series 2 42mm Case Stainless Steel...   
11625    OWC0177  OWC USB-C Dock USB 3.1 / HDMI / Audio / Giga E...   
12995  WDT0302-A  (Open) Western Digital Elements Desktop USB 3....   
13353    OTT0164  OtterBox Defender iPhone Case Black Triple Lay...   
17429    AP20326          Like new - Apple iPhone 7 Plus 256GB Gold   
317      NTE0017        NewerTech NuPower Battery MacBook 13 "white   

                                                    desc    price promo_price  \
11334  Neoprene with reinforced zipper MacBook Air 13...    39.95     179.903   
13426                    4 extra points for Apple Pencil    25.00     249.998   
13172  NAS with 2GB of RAM and two hard drives 4TB (2...   687.69   5.511.792   
1881                      RAM 32GB (4x8GB) Mac Pro 2013.   463.96  

Over 90% of the data in this column is corrupt. There's no point deleting all of these rows, then I would barely have a products table. Instead, as it's only this column that appears to be very untrustworthy, I will delete the column.

In [101]:
products_cl = products_df.drop(columns=["promo_price"])

In [102]:
products_cl.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9992 entries, 0 to 19325
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   sku       9992 non-null   object 
 1   name      9992 non-null   object 
 2   desc      9986 non-null   object 
 3   price     9992 non-null   float64
 4   in_stock  9992 non-null   int64  
 5   type      9946 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 546.4+ KB
